### PR TITLE
yolov9c and yolov8x moving pad, permute and reshape to inside the model

### PR DESCRIPTION
**Pull request by ssinghalTT**
### Problem description
Currently our padding, permute and reshape is on host and is time consuming (order of ms). Moving these to device is fast and helpful. This PR adds these ops to the device before the actual model implementation starts.

### What's changed
The changes are:

- TTNN implementation now includes the padding, permute and reshape.
- Corresponding change to the trace flow if initializing input tensor with the correct mem config shape.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by mbahnasTT — APPROVED** [review]

**Comment by mbahnasTT** [comment]
@ssinghalTT you'd need to run the perf CI tests (and regression tests as well), and update the thresholds of these models accordingly.
The yolo evaluation script will need similar updates, I believe. You can ask @punithsekar to help.

**Comment by ssinghalTT** [comment]
@mbahnasTT I am running the CIs. Ideally should not need any update in the evaluation script.

@punithsekar do you think we need any update in yolo evaluation script?

**Comment by punithsekar** [comment]
@ssinghalTT , Yes, the yolo script needs to be updated for yolov8x eval script(yolov9c is not added yet to yolo eval script). 
It will be similar to a demo execution. The CI's won't show you since yolo evaluation is not added to the CI's.

File: models/demos/yolov8x/tests/yolov8x_test_infra.py [hunk 1]
```diff
@@ -62,19 +62,14 @@ def setup_l1_sharded_input(self, device, torch_input_tensor=None):
...
         # sharded mem config for fold input
-        num_cores = core_grid.x * core_grid.y
-        shard_h = (n * w * h + num_cores - 1) // num_cores
-        grid_size = core_grid
-        grid_coord = ttnn.CoreCoord(grid_size.x - 1, grid_size.y - 1)
-        shard_grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), grid_coord)})
-        shard_spec = ttnn.ShardSpec(shard_grid, (shard_h, 16), ttnn.ShardOrientation.ROW_MAJOR)
-        input_mem_config = ttnn.MemoryConfig(
-            ttnn.types.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.types.BufferType.L1, shard_spec
+        if c == 3:
```
Comment (bbradelTT) [inline models/demos/yolov8x/tests/yolov8x_test_infra.py:65]: This code is in a few places.

This seems like you should change whatever is upstream.
1. Is this because 16 is the minimum number of channels?
2. What happens if c is e.g. 5 or 7?
3. Aren't you handling this in models/demos/yolov8x/tt/ttnn_yolov8x.py?

Depending on what the purpose is, please remove the adjustment, adjust the code, or add a comment explaining why you make this adjustment.

Maybe something like defining a constant somewhere (not necessarily this file), and then adjusting the channel count work. If you want to keep 3, I would suggest a good variable name for it as well.
```
min_channels = 16
...
if c < min_channels:
  c = min_channels
```

**Comment by bbradelTT** [comment]
Please fill in the description.

File: models/demos/yolov8x/tests/yolov8x_test_infra.py (hunk 1 above)
Comment (ssinghalTT) [reply models/demos/yolov8x/tests/yolov8x_test_infra.py:65]: @bbradelTT 
1. Yes.
2. Since this model is image based, c is expected to be 3. 
3. models/demos/yolov8x/tt/ttnn_yolov8x.py handles the changing of input tensor to the correct shape for the conv but models/demos/yolov8x/tests/yolov8x_test_infra.py is needed to get the correct mem config for trace flow as well. We dont want that to be part of the model.

Comment (bbradelTT) [reply models/demos/yolov8x/tests/yolov8x_test_infra.py:65]: Please add comments then, saying that you're adjusting from image based channels to the minimum channels.

**Review by bbradelTT — APPROVED** [review]

**Comment by ssinghalTT** [comment]
@punithsekar I have updated the evaluation script as well for yolov8x. Can you take a look.

**Comment by punithsekar** [comment]
@ssinghalTT , The changes looks good w.r.t yolo evaluation.

**Comment by saichandax** [comment]
APC: [WIP](https://github.com/tenstorrent/tt-metal/actions/runs/16021488393)
Model perf: [pass](https://github.com/tenstorrent/tt-metal/actions/runs/16021462813)
Device perf: [WIP](https://github.com/tenstorrent/tt-metal/actions/runs/16021477125)
Single-card demo: [yolov9c pass](https://github.com/tenstorrent/tt-metal/actions/runs/16021449562)

File: models/demos/yolov8x/runner/performant_runner.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import ttnn
+from models.demos.yolov8x.runner.performant_runner_infra import YOLOv8xPerformanceRunnerInfra
+
+try:
+    pass
+
+    use_signpost = True
+except ModuleNotFoundError:
+    use_signpost = False
+
```
Comment (TT-BrianLiu) [inline models/demos/yolov8x/runner/performant_runner.py]: What does this code do?

File: models/demos/yolov9c/tests/perf/test_e2e_performant.py [hunk 3]
```diff
@@ -56,15 +55,14 @@ def test_e2e_performant(
...
     torch_input_tensor = torch.randn(input_shape, dtype=torch.float32)
-    torch_input_tensor = F.pad(torch_input_tensor, (0, 29), mode="constant", value=0)
-    tt_inputs_host = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
 
     inference_times = []
     for _ in range(10):
         t0 = time.time()
-        _ = performant_runner._execute_yolov9_trace_2cqs_inference(tt_inputs_host)
+        out = performant_runner.run(torch_input_tensor)
+        output = ttnn.to_torch(out[0], dtype=torch.float32)
```
Comment (TT-BrianLiu) [inline models/demos/yolov9c/tests/perf/test_e2e_performant.py]: `output` unused?

File: tests/ttnn/integration_tests/yolov8x/test_yolov8x.py [hunk 4]
```diff
@@ -75,13 +75,17 @@ def test_yolov8x_640(device, input_tensor, use_weights_from_ultralytics):
...
     parameters = custom_preprocessor(device, state_dict, inp_h=inp_h, inp_w=inp_w)
 
-    # pad input channels to 16 to avoid slow interleaved2sharded codepath for 3/8 channels
-    ttnn_input = torch.nn.functional.pad(input_tensor, (0, 0, 0, 0, 0, 13, 0, 0), value=0)
-    ttnn_input = ttnn_input.permute((0, 2, 3, 1))
-    ttnn_input = ttnn_input.reshape(
-        1, 1, ttnn_input.shape[0] * ttnn_input.shape[1] * ttnn_input.shape[2], ttnn_input.shape[3]
+    n, c, h, w = input_tensor.shape
+    if c == 3:
+        c = 16
```
Comment (TT-BrianLiu) [inline tests/ttnn/integration_tests/yolov8x/test_yolov8x.py]: What is this trying to do? Pad up channels to 16 always?

**Review by TT-BrianLiu — APPROVED** [review]
Looks ok, but please address comments

File: models/demos/yolov9c/tests/perf/test_e2e_performant.py (hunk 3 above)
Comment (ssinghalTT) [reply models/demos/yolov9c/tests/perf/test_e2e_performant.py]: @TT-BrianLiu yes, unused but we want to capture the real use case scenario where this variable will be used for post processing. Thus need to run to_torch because the post processing will be done in torch

File: tests/ttnn/integration_tests/yolov8x/test_yolov8x.py (hunk 4 above)
Comment (ssinghalTT) [reply tests/ttnn/integration_tests/yolov8x/test_yolov8x.py]: @TT-BrianLiu yes. Since this model inputs an RGB format image, the channels are **always 3**. But we shard in multiples of 16, so need to pad the inputs to 16 channels.

File: models/demos/yolov8x/runner/performant_runner.py (hunk 2 above)
Comment (ssinghalTT) [reply models/demos/yolov8x/runner/performant_runner.py]: @TT-BrianLiu I don't think it was used. So I removed it now.

**Comment by ssinghalTT** [comment]
@saichandax @punithsekar @sabira-mcw still seeing conflict. Can you fix.
